README.md
# 微信支付 node.js

[![Build Status](https://travis-ci.org/sigoden/wechatpay.svg?branch=master)](https://travis-ci.org/sigoden/wechatpay)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/f019843d36f643378a26840660c10f61)](https://www.codacy.com/app/sigoden/wechatpay?utm_source=github.com&utm_medium=referral&utm_content=sigoden/wechatpay&utm_campaign=Badge_Grade)
[![dependencies Status](https://david-dm.org/sigoden/wechatpay/status.svg)](https://david-dm.org/sigoden/wechatpay)
[![Known Vulnerabilities](https://snyk.io/test/github/sigoden/wechatpay/badge.svg?targetFile=package.json)](https://snyk.io/test/github/sigoden/wechatpay?targetFile=package.json)

> 微信支付 SDK，支持刷卡支付、公众号支付、扫码支付、APP 支付、H5 支付，以及优惠券，红包，企业付款，微信代扣

## 特性

- 使用 Typescript 编写，拥有更安全的类型和更好的代码提示
- 支持所有类型微信支付
- 支持微信代扣
- 提供调试模式

## 开始使用

```js
const path = require("path");
const fs = require("fs");
const {
  PubPay,
  RequestError,
  CommunicationError,
  utils: { getXMLBody }
} = require("@sigodenjs/wechatpay");

const pay = new PubPay({
  appId: "wxb80e5bddb2d804f3",
  key: "6Q9VX4N3WTBM9G9XBL7H1L9PB9ANHLY7",
  mchId: "1434712502",
  pfx: fs.readFileSync(path.resolve(__dirname, "cert.p12"))
});

// 调用统一下单接口
pay
  .unifiedOrder({
    body: "腾讯充值中心-QQ会员充值",
    out_trade_no: "1217752501201407033233368018",
    total_fee: 888,
    spbill_create_ip: "8.8.8.8",
    notify_url: "https://example.com/wechatpay/notify",
    trade_type: "JSAPI",
    openid: "oUpF8uMuAJO_M2pxb1Q9zNjWeS6o"
  })
  .then(res => {
    if (!pay.verifySign(res)) {
      // 签名校验失败
      throw new Error("签名校验失败");
    }
    if (res.result_code === "FAIL") {
      console.log(res.err_code, res.err_code_des);
    } else {
      console.log(res.prepay_id);
    }
  })
  .catch(err => {
    if (err instanceof RequestError) {
      // 请求错误
    } else if (err instanceof CommunicationError) {
      // return_code = FAIL
    }
  });

// 支付结果通知
router.post("/wechatpay/notify", (req, res) => {
  const options = {
    length: req.headers["content-length"],
    limit: "1mb",
    encoding: "utf8"
  };
  getXMLBody(req, options).then(data => {
    pay
      .payNotify(data, async parsedData => {
        if (!pay.verifySign(parsedData)) {
          // 签名校验失败
        }
        if (parsedData.parsedDatault_code === "FAIL") {
          // 业务逻辑失败
        }
        // ...
        return {
          return_code: "SUCCESS",
          return_msg: "OK"
        };
      })
      .then(returnData => {
        res.set("Content-Type", "application/xml; charset=utf-8");
        res.end(returnData);
      });
  });
});
```

## 类和接口

### 业务类
- [AppPay](https://sigoden.github.io/wechatpay/classes/apppay): APP 支付
- [Bank](https://sigoden.github.io/wechatpay/classes/bank): 企业付款
- [Coupon](https://sigoden.github.io/wechatpay/classes/coupon): 代金券
- [LitePay](https://sigoden.github.io/wechatpay/classes/litepay): 小程序支付
- [PubPay](https://sigoden.github.io/wechatpay/classes/pubpay): 公众号支付
- [PubQrPay](https://sigoden.github.io/wechatpay/classes/pubqrpay): 扫码支付
- [PubScanPay](https://sigoden.github.io/wechatpay/classes/pubscanpay): 刷卡支付
- [RedPack](https://sigoden.github.io/wechatpay/classes/redpack): 现金红包
- [WapPay](https://sigoden.github.io/wechatpay/classes/wappay): H5 支付
- [AppEntrust](https://sigoden.github.io/wechatpay/classes/appentrust): APP 代扣
- [LiteEntrust](https://sigoden.github.io/wechatpay/classes/liteentrust): 小程序代扣
- [PubEntrsut](https://sigoden.github.io/wechatpay/classes/pubentrsut): 公众号代扣
- [WapEntrust](https://sigoden.github.io/wechatpay/classes/wapentrust): H5 代扣

### 综合类

- [Pay](https://sigoden.github.io/wechatpay/classes/pay): 组合所有支付类，包含所有支付相关接口
- [Entrust](https://sigoden.github.io/wechatpay/classes/entrust): 组合所有代扣类，包含所有代扣相关接扣

> 如果你只使用了某一项支付，请使用业务类。但如果涉及多种支付，使用综合类更便捷。

### 支付相关接口

- [batchQueryComment](https://sigoden.github.io/wechatpay/classes/pay#batchquerycomment): 拉取订单评价数据
- [downloadBill](https://sigoden.github.io/wechatpay/classes/pay#downloadbill): 下载对账单
- [downloadFundFlow](https://sigoden.github.io/wechatpay/classes/pay#downloadfundflow): 下载资金账单
- [orderQuery](https://sigoden.github.io/wechatpay/classes/pay#orderquery): 查询订单
- [refund](https://sigoden.github.io/wechatpay/classes/pay#refund): 申请退款
- [refundNotify](https://sigoden.github.io/wechatpay/classes/pay#refundnotify): 退款结果通知处理
- [refundQuery](https://sigoden.github.io/wechatpay/classes/pay#refundquery): 查询退款
- [report](https://sigoden.github.io/wechatpay/classes/pay#report): 交易保障
- [closeOrder](https://sigoden.github.io/wechatpay/classes/pay#closeorder): 关闭订单
- [payNotify](https://sigoden.github.io/wechatpay/classes/pay#paynotify): 支付结果通知
- [unifiedOrder](https://sigoden.github.io/wechatpay/classes/pay#unifiedorder): 统一下单
- [authCodeToOpenId](https://sigoden.github.io/wechatpay/classes/pay#authcodetoopenid) - 授权码查询 OpenId
- [microPay](https://sigoden.github.io/wechatpay/classes/pay#micropay) - 提交刷卡支付
- [reverse](https://sigoden.github.io/wechatpay/classes/pay#reverse) - 撤销订单
- [shortURL](https://sigoden.github.io/wechatpay/classes/pay#shorturl) - 短链接转换
- [getPublicKey](https://sigoden.github.io/wechatpay/classes/pay#getpublickey): 获取 RSA 加密公钥
- [getTransferInfo](https://sigoden.github.io/wechatpay/classes/pay#gettransferinfo): 查询企业付款到零钱
- [payBank](https://sigoden.github.io/wechatpay/classes/pay#paybank): 企业付款到银行卡
- [queryBank](https://sigoden.github.io/wechatpay/classes/pay#querybank): 查询企业付款到银行卡
- [transfers](https://sigoden.github.io/wechatpay/classes/pay#transfers): 企业付款到零钱
- [queryCouponStock](https://sigoden.github.io/wechatpay/classes/pay#querycouponstock): 查询代金券批次
- [queryCouponsInfo](https://sigoden.github.io/wechatpay/classes/pay#querycouponsinfo): 查询代金券信息
- [sendCoupon](https://sigoden.github.io/wechatpay/classes/pay#sendcoupon): 发放代金券
- [getHbInfo](https://sigoden.github.io/wechatpay/classes/pay#gethbinfo): 查询红包记录
- [sendGroupGroupRedPack](https://sigoden.github.io/wechatpay/classes/pay#sendgroupgroupredpack): 发放裂变红包
- [sendRedPack](https://sigoden.github.io/wechatpay/classes/pay#sendredpack): 发放普通红包

### 代扣相关接口

- [contractNotify](https://sigoden.github.io/wechatpay/classes/entrust#contractnotify): 签约，解约结果通知
- [contractOrder](https://sigoden.github.io/wechatpay/classes/entrust#contractorder): 支付中签约
- [deleteContract](https://sigoden.github.io/wechatpay/classes/entrust#deletecontract): 申请解约
- [entrust](https://sigoden.github.io/wechatpay/classes/entrust#entrust): 纯签约
- [papOrderQuery](https://sigoden.github.io/wechatpay/classes/entrust#paporderquery): 查询订单
- [papPayApply](https://sigoden.github.io/wechatpay/classes/entrust#pappayapply): 申请扣款
- [papPayNotify](https://sigoden.github.io/wechatpay/classes/entrust#pappaynotify): 扣款结果通知
- [queryContract](https://sigoden.github.io/wechatpay/classes/entrust#querycontract): 查询签约关系
- [downloadBill](https://sigoden.github.io/wechatpay/classes/entrust#downloadbill): 下载对账单
- [refund](https://sigoden.github.io/wechatpay/classes/entrust#refund): 申请退款
- [refundQuery](https://sigoden.github.io/wechatpay/classes/entrust#refundquery): 查询退款

## 许可证

Copyright (c) 2018 sigoden

Licensed under the MIT license.


CHANGELOG.md
## v1.1.0

- 添加微信代扣

## v2.0.0

- 使用 typescript 重构

src/pays/Pay.ts
import { fetch } from "../fetch";
import * as types from "../types";
import { Bank } from "./Bank";
import PayBase from "./core/PayBase";
import PayBaseX from "./core/PayBaseX";
import { Coupon } from "./Coupon";
import { PubScanPay } from "./PubScanPay";
import { RedPack } from "./RedPack";

const REPORT_BASE = "/payitil/report";
const UNIFIED_ORDER_BASE = "/pay/unifiedorder";

/**
 * 总和类，糅合所有接口
 *
 * ```
 * const pay = new Pay({
 *   appId: "wxb80e5bddb2d804f3",
 *   key: "6Q9VX4N3WTBM9G9XBL7H1L9PB9ANHLY7",
 *   mchId: "1434712502",
 *   pfx: fs.readFileSync(path.resolve(__dirname, "cert.p12"))
 * });
 * ```
 */
export class Pay extends PayBase {
  private readonly payBaseX: PayBaseX;
  private readonly pubScanPay: PubScanPay;
  private readonly bank: Bank;
  private readonly coupon: Coupon;
  private readonly redPack: RedPack;
  constructor(options: types.ConstructorOptions) {
    super(options);
    this.payBaseX = new PayBaseX(options);
    this.pubScanPay = new PubScanPay(options);
    this.bank = new Bank(options);
    this.coupon = new Coupon(options);
    this.redPack = new RedPack(options);
  }

  /**
   * 关闭订单
   *
   * ```
   * pay.closeOrder({
   *   out_trade_no: '120061098828009406'
   * })
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/jsapi.php?chapter=9_3}
   */
  public async closeOrder(options: types.CloseOrderOptions) {
    return this.payBaseX.closeOrder(options);
  }

  /**
   * 交易保障
   *
   * ```
   * pay.report({
   *   interface_url: "https://api.mch.weixin.qq.com/pay/unifiedorder",
   *   execute_time: 1000,
   *   return_code: "SUCCESS",
   *   return_msg: "OK",
   *   result_code: "SUCCESS",
   *   user_ip: "8.8.8.8"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/jsapi.php?chapter=9_8&index=9} or
   *   {@link https://pay.weixin.qq.com/wiki/doc/api/micropay.php?chapter=9_10&index=1}
   */
  public async report(
    options: types.ReportOptionsGeneral | types.ReportOptionsPubScan
  ) {
    const url = this.completeURL(REPORT_BASE);
    const extra = await this.createFetchOptions(url);
    return fetch<
      types.ReportOptionsGeneral | types.ReportOptionsPubScan,
      types.ReportSuccess,
      types.ReportFail
    >(options, extra);
  }

  /**
   * 支付结果通知
   *
   * ```
   * router.post("/wechatpay/notify/refund", (req, res) => {
   *   getXMLBody(req, options).then(data => {
   *     pay
   *       .payNotify(data, async parsedData => {
   *         // ...
   *         return {
   *           return_code: "SUCCESS",
   *           return_msg: "OK"
   *         };
   *       })
   *       .then(returnData => {
   *         res.set("Content-Type", "application/xml; charset=utf-8");
   *         res.end(returnData);
   *       });
   *   });
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/jsapi.php?chapter=9_7&index=8}
   */
  public async payNotify(
    info: types.PayNotifySuccess | types.PayNotifyFail,
    handler: types.PayNotifyHandler
  ) {
    return this.payBaseX.payNotify(info, handler);
  }

  /**
   * 统一下单
   *
   * ```
   * pay.unifiedOrder({
   *   trade_type: "APP",
   *   body: "腾讯充值中心-QQ会员充值",
   *   out_trade_no: "1217752501201407033233368018",
   *   total_fee: 888,
   *   spbill_create_ip: "8.8.8.8",
   *   notify_url: "https://example.com/wechatpay/notify"
   * });
   * ```
   */
  public async unifiedOrder(options: types.UnifiedOrderOptionsGeneral) {
    const url = this.completeURL(UNIFIED_ORDER_BASE);
    const extra = await this.createFetchOptions(url);
    return fetch<
      types.UnifiedOrderOptionsGeneral,
      types.UnifiedOrderSuccess,
      types.UnifiedOrderFail
    >(options, extra);
  }

  /**
   * 提交刷卡支付
   *
   * ```
   * pay.microPay({
   *   device_info: "013467007045764",
   *   body: "深圳腾大- QQ公仔",
   *   out_trade_no: "1217752501201407033233368018",
   *   total_fee: 888,
   *   spbill_create_ip: "8.8.8.8",
   *   auth_code: "120061098828009406"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/micropay.php?chapter=9_10&index=1}
   */
  public async microPay(options: types.MicroPayOptions) {
    return this.pubScanPay.microPay(options);
  }

  /**
   * 撤销订单
   *
   * ```
   * pay.reverse({
   *   transaction_id: '1009660380201506130728806387'
   * })
   *
   * pay.reverse({
   *   out_trade_no: '120061098828009406'
   * })
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/micropay.php?chapter=9_11&index=3}
   */
  public async reverse(options: types.ReverseOptions) {
    return this.pubScanPay.reverse(options);
  }
  /**
   * 短链接转换
   *
   * ```
   * pay.shortURL({
   *   long_url:
   *     "weixin://wxpay/bizpayurl?sign=XXX&appid=XX&mch_id=XX&product_id=XX&time_stamp=XX&nonce_str=XX"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/micropay.php?chapter=9_9&index=9}
   */
  public async shortURL(options: types.ShortURLOptions) {
    return this.pubScanPay.shortURL(options);
  }

  /**
   * 授权码查询 OpenId
   *
   * ```
   * pay.authCodeToOpenId({
   *   auth_code: "120061098828009406"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/micropay.php?chapter=9_13&index=10}
   */
  public async authCodeToOpenId(options: types.AuthCodeToOpenIdOptions) {
    return this.pubScanPay.authCodeToOpenId(options);
  }

  /**
   * 企业付款到零钱
   *
   * ```
   * pay.transfers({
   *   partner_trade_no: "10000098201411111234567890",
   *   openid: "oxTWIuGaIt6gTKsQRLau2M0yL16E",
   *   check_name: "FORCE_CHECK",
   *   amount: 10099,
   *   desc: "理赔",
   *   spbill_create_ip: "192.168.0.1"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/mch_pay.php?chapter=14_2}
   */
  public async transfers(options: types.TransfersOptions) {
    return this.bank.transfers(options);
  }

  /**
   * 查询企业付款到零钱
   *
   * ```
   * pay.getTransferInfo({
   *   partner_trade_no: "10000098201411111234567890"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/mch_pay.php?chapter=14_3}
   */
  public async getTransferInfo(options: types.GetTransferInfoOptions) {
    return this.bank.getTransferInfo(options);
  }

  /**
   * 企业付款到银行卡
   *
   * ```
   * pay.payBank({
   *   partner_trade_no: "1212121221227",
   *   enc_bank_no: utils.rsa(pemKey, "6225760017946512"),
   *   enc_true_name: utils.rsa(pemKey, "王小王"),
   *   bank_code: "1001",
   *   amount: 100,
   *   desc: "理财"
   * });
   * // pemKey 通过 `pay.getPublicKey` 获取
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/mch_pay.php?chapter=24_2}
   */
  public async payBank(options: types.PayBankOptions) {
    return this.bank.payBank(options);
  }

  /**
   * 查询企业付款到银行卡
   *
   * ```
   * pay.queryBank({
   *   partner_trade_no: "1212121221227"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/mch_pay.php?chapter=24_3}
   */
  public async queryBank(options: types.QueryBankOptions) {
    return this.bank.queryBank(options);
  }

  /**
   * 获取RSA加密公钥
   *
   * ```
   * pay.getPublicKey({
   *  sign_type: SignType.MD5
   * });
   * ```
   */
  public async getPublicKey(options: types.GetPublicKeyOptions) {
    return this.bank.getPublicKey(options);
  }

  /**
   * 发放代金券
   *
   * ```
   * pay.sendCoupon({
   *   coupon_stock_id: "1757",
   *   openid_count: 1,
   *   partner_trade_no: "1000009820141203515766",
   *   openid: "onqOjjrXT-776SpHnfexGm1_P7iE"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/sp_coupon.php?chapter=12_3&index=4}
   */
  public async sendCoupon(options: types.SendCouponOptions) {
    return this.coupon.sendCoupon(options);
  }

  /**
   * 查询代金券批次
   *
   * ```
   * pay.queryCouponStock({
   *   coupon_stock_id: "1757"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/sp_coupon.php?chapter=12_4&index=5}
   */
  public async queryCouponStock(options: types.QueryCouponStockOptions) {
    return this.coupon.queryCouponStock(options);
  }

  /**
   * 查询代金券信息
   *
   * ```
   * pay.queryCouponsInfo({
   *   coupon_id: "1565",
   *   openid: "onqOjjrXT-776SpHnfexGm1_P7iE",
   *   stock_id: "58818"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/sp_coupon.php?chapter=12_5&index=6}
   */
  public async queryCouponsInfo(options: types.QueryCouponsInfoOptions) {
    return this.coupon.queryCouponsInfo(options);
  }

  /**
   * 发放普通红包
   *
   * ```
   * pay.sendRedPack({
   *   mch_billno: "10000098201411111234567890",
   *   send_name: "天虹百货",
   *   re_openid: "oxTWIuGaIt6gTKsQRLau2M0yL16E",
   *   total_amount: 1000,
   *   total_num: 1,
   *   wishing: "红包祝福语",
   *   client_ip: "192.168.0.1",
   *   act_name: "猜灯谜抢红包活动",
   *   remark: "猜越多得越多，快来抢！"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/cash_coupon.php?chapter=13_4&index=3}
   */
  public async sendRedPack(options: types.SendRedPackOptions) {
    return this.redPack.sendRedPack(options);
  }

  /**
   * 发放裂变红包
   *
   * ```
   * pay.sendGroupGroupRedPack({
   *   mch_billno: "10000098201411111234567890",
   *   send_name: "天虹百货",
   *   re_openid: "oxTWIuGaIt6gTKsQRLau2M0yL16E",
   *   total_amount: 1000,
   *   total_num: 1,
   *   wishing: "红包祝福语",
   *   act_name: "猜灯谜抢红包活动",
   *   remark: "猜越多得越多，快来抢！"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/cash_coupon.php?chapter=13_5&index=4}
   */
  public async sendGroupGroupRedPack(options: types.SendGroupRedPackOptions) {
    return this.redPack.sendGroupGroupRedPack(options);
  }

  /**
   * 查询红包记录
   *
   * ```
   * pay.getHbInfo({
   *   mch_billno: "10000098201411111234567890"
   * });
   * ```
   * @see {@link https://pay.weixin.qq.com/wiki/doc/api/tools/cash_coupon.php?chapter=13_6&index=5}
   */
  public async getHbInfo(options: types.GetHbInfoOptions) {
    return this.redPack.getHbInfo(options);
  }
}
